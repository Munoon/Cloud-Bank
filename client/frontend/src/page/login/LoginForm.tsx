import React, {useCallback} from 'react';
import {connect, ConnectedProps, useSelector} from 'react-redux';
import {useForm} from 'react-hook-form';

import {login} from './middleware';
import {readProperty} from '../../utils/globalUtils';
import {loginShowPassword, LoginState, LoginStoreStateType, LoginStoreType} from "./store";

const connector = connect(null, { login });

export interface LoginFormData {
    username: string;
    password: string;
}

const LoginForm = ({ login }: ConnectedProps<typeof connector>) => {
    const { handleSubmit, register, errors } = useForm();
    const onSubmit = useCallback((data: LoginFormData) => login(data), []);
    const showPassword = useSelector((store: LoginStoreType) => store.login.showPassword);

    return (
        <form action={`${readProperty('base_url')}/login`} method='post'
              onSubmit={handleSubmit(onSubmit)} className='object-center font-mono mt-3'>
            <FailMessage />
            <input type='hidden' name='_csrf' value={readProperty('tokens:csrf')} />
            <FormInput placeholder='Username' name='username' ref={register({ required: true })} />
            <FormInput type={showPassword ? 'text' : 'password'}
                       placeholder='Password' name='password' ref={register({ required: true })} />
            <SupportButtons />
            <SubmitButton disabled={Object.entries(errors).length > 0} />
        </form>
    );
}

const SubmitButton = ({ disabled }: { disabled: boolean }) => {
    const loginState = useSelector((store: LoginStoreType) => store.login.state);
    const shouldDisabled = disabled || loginState === LoginState.LOADING;
    const loginSuccess = loginState === LoginState.LOGGED_IN;
    return (
        <button type="submit"
                className={(loginSuccess ? 'bg-success' : 'bg-primary')
                    + ' disabled:bg-disabled mt-10 p-1 transition rounded-full'
                    + ' w-full text-center text-white focus:outline-none'}
                disabled={shouldDisabled}>Sign in</button>
    );
}

const failMessages: Record<string, string> = {
    'bad_credentials': 'Incorrect login or password!'
};

const FailMessage = () => {
    const login: LoginStoreStateType = useSelector((store: LoginStoreType) => store.login);
    const loginFailed = login.state === LoginState.ERROR;
    if (!loginFailed) {
        return null;
    }
    const errorCode = login.errorData.errorCode;
    if (errorCode && failMessages[errorCode] !== undefined) {
        return <div>{failMessages[errorCode]}</div>;
    }
    return <div>{login.errorData.errorMessage}</div>
}

const SupportButtons = () => (
    <div className='flex items-center justify-between mt-1'>
        <ShowPasswordButton />
        <ForgotPasswordButton />
    </div>
);

const showPasswordButtonConnector = connect(
    (state: LoginStoreType) => ({ showPassword: state.login.showPassword }),
    { loginShowPassword }
);
const ShowPasswordButton = showPasswordButtonConnector((props: ConnectedProps<typeof showPasswordButtonConnector>) => (
    <div className='flex items-center'>
        <input onChange={e => props.loginShowPassword(e.target.checked)}
               checked={props.showPassword}
               id="showPassword"
               type="checkbox"
               className='h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded' />
        <label htmlFor="showPassword"
               className="ml-2 block text-sm text-gray-900 select-none">Show password</label>
    </div>
));

const ForgotPasswordButton = () => (
    <div className="text-sm">
        <a href="#" className='font-medium text-indigo-600 hover:text-indigo-500 select-none'>Forgot password?</a>
    </div>
);

interface FormInputProps {
    type?: string;
    name?: string;
    placeholder?: string;
}

const FormInput = React.forwardRef<HTMLInputElement, FormInputProps>((props, ref)  => (
    <input type={props.type} name={props.name} placeholder={props.placeholder} ref={ref}
           className="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary" />
));

export default connector(LoginForm);