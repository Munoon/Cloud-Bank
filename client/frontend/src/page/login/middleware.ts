import {Dispatch} from 'redux';

import {LoginFormData} from "./LoginForm";
import {loginErrorAction, loginLoadingAction, loginLoggedInAction} from './reducer';
import {readProperty} from "../../utils/globalUtils";
import fetcher from "../../utils/fetcher";

export const login = (data: LoginFormData) => async (dispatch: Dispatch) => {
    dispatch(loginLoadingAction());
    const url = `${readProperty('base_url')}/login`;
    const params = {
        username: data.username,
        password: data.password
    };
    try {
        const info = await fetcher<SuccessAuthenticationInfo>(url, { method: 'POST', params, redirect: 'manual' });
        dispatch(loginLoggedInAction());
        location.href = info.redirectUrl;
    } catch (e) {
        const failAuthenticationInfo = e.response as FailAuthenticationInfo;
        dispatch(loginErrorAction(failAuthenticationInfo));
    }
}

interface SuccessAuthenticationInfo {
    status: string;
    redirectUrl: string;
}

interface FailAuthenticationInfo {
    status: string;
    errorCode?: string;
    errorMessage?: string;
}