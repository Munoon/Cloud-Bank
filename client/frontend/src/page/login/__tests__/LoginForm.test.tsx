import React from "react";
import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import {Provider} from "react-redux";

import LoginForm, {
    FailMessage,
    ForgotPasswordButton,
    FormInput,
    ShowPasswordButton,
    SubmitButton,
    SupportButtons
} from "../LoginForm";
import mockStore from "./utils/mockStore";
import {loginShowPassword, LoginState} from "../reducer";
import {login} from '../middleware';

jest.mock('../middleware', () => ({
    login: jest.fn(() => async () => {})
}));

describe('Login form test', () => {
    describe('Functional test', () => {
        it('button status disabled because of invalid inputs', async () => {
            render(
                <Provider store={mockStore()}>
                    <LoginForm />
                </Provider>
            );

            const button = screen.getByRole('button');
            fireEvent.submit(button);


            await waitFor(() => {});
            expect(button).toHaveAttribute('disabled');
        });

        it('button status disabled because of logged in', () => {
            render(
                <Provider store={mockStore({ login: { state: LoginState.LOGGED_IN } })}>
                    <LoginForm />
                </Provider>
            );

            const button = screen.getByRole('button');
            expect(button).toHaveAttribute('disabled');
            expect(button).toHaveClass('bg-success');
        });

        it('button status clickable because input valid', async () => {
            render(
                <Provider store={mockStore()}>
                    <LoginForm />
                </Provider>
            );

            const usernameInput = screen.getByPlaceholderText('Username');
            fireEvent.change(usernameInput, { target: { value: 'test' } });
            expect(usernameInput).toHaveValue('test');

            const passwordInput = screen.getByPlaceholderText('Password');
            fireEvent.change(passwordInput, { target: { value: 'test' } });
            expect(passwordInput).toHaveValue('test');

            const button = screen.getByRole('button');
            fireEvent.submit(button);

            await waitFor(() => {
                expect(button).not.toHaveAttribute('disabled');
            });
        });

        it('button status clickable on startup', () => {
            const store = mockStore();
            render(
                <Provider store={store}>
                    <LoginForm />
                </Provider>
            );

            const button = screen.getByRole('button');
            expect(button).not.toHaveAttribute('disabled');
        });

        it('show password', () => {
            render(
                <Provider store={mockStore({ login: { showPassword: true } })}>
                    <LoginForm />
                </Provider>
            );

            const passwordInput = screen.getByPlaceholderText('Password');
            expect(passwordInput).toHaveAttribute('type', 'text');
        });

        it('hide password', () => {
            render(
                <Provider store={mockStore({ login: { showPassword: false } })}>
                    <LoginForm />
                </Provider>
            );

            const passwordInput = screen.getByPlaceholderText('Password')
            expect(passwordInput).toHaveAttribute('type', 'password');
        });

        it('send form', async () => {
            const store = mockStore();
            render(
                <Provider store={store}>
                    <LoginForm />
                </Provider>
            );

            const usernameInput = screen.getByPlaceholderText('Username');
            fireEvent.change(usernameInput, { target: { value: 'test' } });
            expect(usernameInput).toHaveValue('test');

            const passwordInput = screen.getByPlaceholderText('Password');
            fireEvent.change(passwordInput, { target: { value: 'test' } });
            expect(passwordInput).toHaveValue('test');

            const button = screen.getByRole('button');
            fireEvent.submit(button);

            await waitFor(() => {
                expect(login).toHaveBeenLastCalledWith({ username: 'test', password: 'test' });
            });
        });
    });

    it('Snapshot test', () => {
        const component = render(
            <Provider store={mockStore()}>
                <LoginForm />
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});

describe('Submit button component test', () => {
    it('Disabled by hand', () => {
        render(
            <Provider store={mockStore()}>
                <SubmitButton disabled={true} />
            </Provider>
        );

        const button = screen.getByRole('button');
        expect(button).toHaveAttribute('disabled');
    });

    it('Disabled because loading', () => {
        render(
            <Provider store={mockStore({ login: { state: LoginState.LOADING } })}>
                <SubmitButton disabled={false} />
            </Provider>
        );

        const button = screen.getByRole('button');
        expect(button).toHaveAttribute('disabled');
    });

    it('Disabled because logged in', () => {
        render(
            <Provider store={mockStore({ login: { state: LoginState.LOGGED_IN } })}>
                <SubmitButton disabled={false} />
            </Provider>
        );

        const button = screen.getByRole('button');
        expect(button).toHaveAttribute('disabled');
    });

    it('Have default background color', () => {
        render(
            <Provider store={mockStore()}>
                <SubmitButton disabled={false} />
            </Provider>
        );

        const button = screen.getByRole('button');
        expect(button).toHaveClass('bg-primary');
        expect(button).not.toHaveClass('bg-success');
    });

    it('Have success background color on loading', () => {
        render(
            <Provider store={mockStore({ login: { state: LoginState.LOGGED_IN } })}>
                <SubmitButton disabled={false} />
            </Provider>
        );

        const button = screen.getByRole('button');
        expect(button).not.toHaveClass('bg-primary');
        expect(button).toHaveClass('bg-success');
    });

    it('Snapshot test', () => {
        const component = render(
            <Provider store={mockStore()}>
                <SubmitButton disabled={false} />
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});

describe('Fail message component test', () => {
    it('No message test', () => {
        render(
            <Provider store={mockStore()}>
                <FailMessage />
            </Provider>
        );

        const component = screen.queryByTestId('login_form.error_message');
        expect(component).toBeNull();
    });

    it('Message by error code', () => {
        const storeModification = {
            login: {
                state: LoginState.ERROR,
                errorData: {
                    errorCode: 'bad_credentials',
                    errorMessage: 'test error'
                }
            }
        };
        render(
            <Provider store={mockStore(storeModification)}>
                <FailMessage />
            </Provider>
        );

        const component = screen.queryByTestId('login_form.error_message');
        expect(component).toHaveTextContent('Incorrect login or password!');
    });

    it('Message from server', () => {
        const storeModification = {
            login: {
                state: LoginState.ERROR,
                errorData: { errorMessage: 'test error' }
            }
        };
        render(
            <Provider store={mockStore(storeModification)}>
                <FailMessage />
            </Provider>
        );

        const component = screen.queryByTestId('login_form.error_message');
        expect(component).toHaveTextContent('test error');
    });

    it('Snapshot test', () => {
        const storeModification = {
            login: {
                state: LoginState.ERROR,
                errorData: { errorMessage: 'test error' }
            }
        };
        const component = render(
            <Provider store={mockStore(storeModification)}>
                <FailMessage />
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});

describe('Support buttons component test', () => {
    it('Snapshot test', () => {
        const component = render(
            <Provider store={mockStore()}>
                <SupportButtons />
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});

describe('Forgot password button component test', () => {
    it('Snapshot test', () => {
        const component = render(<ForgotPasswordButton />);
        expect(component).toMatchSnapshot();
    });
});

describe('Show password checkbox component test', () => {
    it('Show password', () => {
        render(
            <Provider store={mockStore({ login: { showPassword: true } })}>
                <ShowPasswordButton />
            </Provider>
        );

        const element = screen.getByRole('checkbox');
        expect(element).toHaveAttribute('checked');
    });

    it('Hide password', () => {
        render(
            <Provider store={mockStore()}>
                <ShowPasswordButton />
            </Provider>
        );

        const element = screen.getByRole('checkbox');
        expect(element).not.toHaveAttribute('checked');
    });

    it('Click to show password', () => {
        const store = mockStore();
        render(
            <Provider store={store}>
                <ShowPasswordButton />
            </Provider>
        );

        const element = screen.getByRole('checkbox');
        fireEvent.click(element);

        const actions = store.getActions();
        const lastAction = actions[actions.length - 1];
        expect(lastAction).toEqual(loginShowPassword(true));
    });

    it('Click to hide password', () => {
        const store = mockStore({ login: { showPassword: true } });
        render(
            <Provider store={store}>
                <ShowPasswordButton />
            </Provider>
        );

        const element = screen.getByRole('checkbox');
        fireEvent.click(element);

        const actions = store.getActions();
        const lastAction = actions[actions.length - 1];
        expect(lastAction).toEqual(loginShowPassword(false));
    });

    it('Snapshot test', () => {
        const component = render(
            <Provider store={mockStore({ login: { showPassword: true } })}>
                <ShowPasswordButton />
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});

describe('Form input test', () => {
    it('Functional test', () => {
        render(
            <FormInput type='password' name='testName' placeholder='testPlaceholder' />
        );

        const element = screen.getByPlaceholderText('testPlaceholder');
        expect(element).toHaveAttribute('type', 'password');
        expect(element).toHaveAttribute('name', 'testName');
        expect(element).toHaveAttribute('placeholder', 'testPlaceholder');
    });

    it('Snapshot test', () => {
        const component = render(
            <FormInput type='password' name='testName' placeholder='testPlaceholder' />
        );
        expect(component).toMatchSnapshot();
    });
});