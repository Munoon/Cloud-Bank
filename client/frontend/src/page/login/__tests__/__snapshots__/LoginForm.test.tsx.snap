// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fail message component test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-red-500 p-3 w-full text-white"
        data-testid="login_form.error_message"
      >
        test error
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-red-500 p-3 w-full text-white"
      data-testid="login_form.error_message"
    >
      test error
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Forgot password button component test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="text-sm"
      >
        <a
          class="font-medium text-indigo-600 hover:text-indigo-500 select-none"
          href="#"
        >
          Forgot password?
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="text-sm"
    >
      <a
        class="font-medium text-indigo-600 hover:text-indigo-500 select-none"
        href="#"
      >
        Forgot password?
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Form input test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <input
        class="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary"
        name="testName"
        placeholder="testPlaceholder"
        type="password"
      />
    </div>
  </body>,
  "container": <div>
    <input
      class="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary"
      name="testName"
      placeholder="testPlaceholder"
      type="password"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login form test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        action="http://localhost/login"
        class="object-center font-mono mt-3"
        method="post"
      >
        <input
          name="_csrf"
          type="hidden"
          value="MOCK_CSRF_TOKEN"
        />
        <input
          class="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary"
          name="username"
          placeholder="Username"
        />
        <input
          class="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary"
          name="password"
          placeholder="Password"
          type="password"
        />
        <div
          class="flex items-center justify-between mt-1"
        >
          <div
            class="flex items-center"
          >
            <input
              class="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
              id="showPassword"
              type="checkbox"
            />
            <label
              class="ml-2 block text-sm text-gray-900 select-none"
              for="showPassword"
            >
              Show password
            </label>
          </div>
          <div
            class="text-sm"
          >
            <a
              class="font-medium text-indigo-600 hover:text-indigo-500 select-none"
              href="#"
            >
              Forgot password?
            </a>
          </div>
        </div>
        <button
          class="bg-primary disabled:bg-disabled mt-10 p-1 transition rounded-full w-full text-center text-white focus:outline-none"
          type="submit"
        >
          Sign in
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      action="http://localhost/login"
      class="object-center font-mono mt-3"
      method="post"
    >
      <input
        name="_csrf"
        type="hidden"
        value="MOCK_CSRF_TOKEN"
      />
      <input
        class="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary"
        name="username"
        placeholder="Username"
      />
      <input
        class="mt-10 bg-primary p-1 rounded-full w-full text-center text-white outline-none placeholder-secondary"
        name="password"
        placeholder="Password"
        type="password"
      />
      <div
        class="flex items-center justify-between mt-1"
      >
        <div
          class="flex items-center"
        >
          <input
            class="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
            id="showPassword"
            type="checkbox"
          />
          <label
            class="ml-2 block text-sm text-gray-900 select-none"
            for="showPassword"
          >
            Show password
          </label>
        </div>
        <div
          class="text-sm"
        >
          <a
            class="font-medium text-indigo-600 hover:text-indigo-500 select-none"
            href="#"
          >
            Forgot password?
          </a>
        </div>
      </div>
      <button
        class="bg-primary disabled:bg-disabled mt-10 p-1 transition rounded-full w-full text-center text-white focus:outline-none"
        type="submit"
      >
        Sign in
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Show password checkbox component test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center"
      >
        <input
          checked=""
          class="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
          id="showPassword"
          type="checkbox"
        />
        <label
          class="ml-2 block text-sm text-gray-900 select-none"
          for="showPassword"
        >
          Show password
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center"
    >
      <input
        checked=""
        class="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
        id="showPassword"
        type="checkbox"
      />
      <label
        class="ml-2 block text-sm text-gray-900 select-none"
        for="showPassword"
      >
        Show password
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Submit button component test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="bg-primary disabled:bg-disabled mt-10 p-1 transition rounded-full w-full text-center text-white focus:outline-none"
        type="submit"
      >
        Sign in
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="bg-primary disabled:bg-disabled mt-10 p-1 transition rounded-full w-full text-center text-white focus:outline-none"
      type="submit"
    >
      Sign in
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Support buttons component test Snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center justify-between mt-1"
      >
        <div
          class="flex items-center"
        >
          <input
            class="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
            id="showPassword"
            type="checkbox"
          />
          <label
            class="ml-2 block text-sm text-gray-900 select-none"
            for="showPassword"
          >
            Show password
          </label>
        </div>
        <div
          class="text-sm"
        >
          <a
            class="font-medium text-indigo-600 hover:text-indigo-500 select-none"
            href="#"
          >
            Forgot password?
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center justify-between mt-1"
    >
      <div
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
          id="showPassword"
          type="checkbox"
        />
        <label
          class="ml-2 block text-sm text-gray-900 select-none"
          for="showPassword"
        >
          Show password
        </label>
      </div>
      <div
        class="text-sm"
      >
        <a
          class="font-medium text-indigo-600 hover:text-indigo-500 select-none"
          href="#"
        >
          Forgot password?
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
